### ISSUE TITLE
OS.execute does not work with windows system commands and has no way of setting the working path

### ISSUE THREAD

**Issue by blurymind** [godotengine/godot#5708]
OS.execute does not seem to work with windows' 'start' command - used to open a file with the default application assigned to its file type.
It works in CMD terminal - but not when i try to do it through godot.

https://gist.github.com/blurymind/02325f99edd446009834cbddecf02212

Other commands that have been identified to not work with OS.execute are 
dir  echo  cat   and some other system commands.

A possible connection to this:
http://www.frankworsley.com/blog/2007/9/24/open-a-file-in-the-default-application-using-the-windows-command-line-without-jdic

Fortunately there is a workaround for this. Using OS.shell_open ('path-to-target') instead would open a CMD terminal and run the command.

I reported this on FB - but was adviced to also file it here. I hope it helps :)
https://www.facebook.com/groups/godotengine/permalink/803676253102304/?comment_id=804178909718705&reply_comment_id=804804472989482&ref=notif&notif_t=group_comment&notif_id=1468474364320939

**Comment by punto-** [comment]
The method "OS.shell_open()" does the equivalent of "start", and is
portable, it'll work on all other OSs

**Comment by blurymind** [comment]
@punto-  Yes I know - That is why I mentioned it. :)

The feedback on facebook was that I should post it to the issue tracker anyways.

**Comment by blurymind** [comment]
Sorry I wrote a wrong syntax in above post.

Is should be  OS.shell_open('path_to_file_or_folder')

OS.shell can also be used to start other system commands I think - but their output can not be easily fed to godot to my knowledge.

**Comment by blurymind** [comment]
Hi Sorry to wake this old thingup, but is there a way to set OS.execute's working path?
I am trying to get it to trigger zip.exe to zip a folder with relative recursive paths

**Comment by blurymind** [comment]
How the hell does one set current path of execution of OS.execute??

**Comment by eon-s** [comment]
Maybe `start /d c:\working_directory excutable.exe` in the case of windows?
https://technet.microsoft.com/en-us/library/cc770297(v=ws.11).aspx

Could be nice to have it as a second parameter for every (or most) platforms.

**Comment by blurymind** [comment]
@eon-s Thank you. Does this work on windows 10 as well?
Also how do you do this in gdscript?
OS.execute has trouble with windows system commands- it doesn't recognize them at all
[try.zip](https://github.com/godotengine/godot/files/1567354/try.zip)

I am attaching a file you can try to do it with.
The folder contains:

> zip.exe
> bzip2.dll
> data.smix
> 0 (folder)

The goal is to run zip exe like this:
`zip.exe -r result.zip 0 data.smix`

And inside the resulting zip file's root you have:
> 0 (and it's contents - not full path)
> data.smix

In windows cmd you can do
`start /d C:yourWorkingPath\$tmp--test zip.exe -r result.zip 0 data.smix`

Os.execute() will probably fail to recognize the start command. I am going to try it and report on it

Basically the problem now is more than one issue:
- OS.execute() Has no way of setting working path - this cripples a lot of applications that need that as a part of their design in order to execute. Zip.exe is a prime example
- OS.execute() has no support for system commands, which again stops us from setting path by parsing it to the command line
- Godot's zipping and unzipping features are not exposed to gdscript

**Comment by blurymind** [comment]
I will also try with OS.shell_open('start /d C:\yourWorkingPath\$tmp--test zip.exe -r result.zip 0 data.smix') , although I am somewhat skeptical

**Comment by blurymind** [comment]
```
OS.shell_open('start /d F:\\stuff\\godot\\godot\\SPRITEmixer\\$tmp--test zip.exe -r result.zip
data.smix')
```
Fails miserably - nothing happens and what's worse - I get no output about the error

From my perspective Godot can't execute anything with a custom working path

**Comment by brainsick** [comment]
@blurymind What are the origins of this zip.exe?  Might it have a command line option to specify a 'base' directory?

**Comment by blurymind** [comment]
@brainsick  good thinking. I tried to find one, but I couldnt. -d seems to be the closest, but that doesnt work - its just for setting the tmp dir

I am using the binary version of
http://gnuwin32.sourceforge.net/packages/zip.htm
Please let me know if you find a way :)

**Comment by brainsick** [comment]
@blurymind I looked at the zip docs and didn't see anything appropriate.  Just to confirm, you need this to work on Windows and you don't care about portability?  I don't have access to Windows, so I don't think I can help, sorry.  Good luck.

**Comment by blurymind** [comment]
@brainsick I would like to have portability. That is why I chose zip- its open source and works on linux,windows and mac

**Comment by brainsick** [comment]
@blurymind I don't mean to detract from your reported issue.  I don't see a method of accomplishing this in the currently exposed API either.

I'm just trying to help with your use case.  Can you post the exact `OS.execute` syntax of your failed `start` command attempt?  This should be a Linux equivalent, which does work for me:

```
	# *nix solution
	pid = OS.execute('/bin/sh', ['-c', 'cd /tmp && echo \\`pwd\\`'], true, output)
	print('pid: ', pid, ' output: ', output)
```

**Comment by blurymind** [comment]
@brainsick thank you for the help so far. I tried your last suggestion and that did not work either.

It would be very nice if OS.execute has a way of setting paths.

In my case I simply abandoned zip.exe and replaced it with 7za.exe - which is half a mb when you strip the dll files that come with it. Zip has a pain in the arse design that godot couldnt handle due to this limitation. I got tired of it and moved onto command line 7zip instead

**Comment by brainsick** [comment]
@blurymind I'm sensing that you don't want to troubleshoot this further.  Should this issue be closed?  I'd recommend opening new issues for the API enhancements so they can be discussed on their own merits without the troubleshooting noise.

**Comment by blurymind** [comment]
I am going to give it a chance a bit more when I go back home :) But for my app, ultimatelly my decision was that zip.exe is not a good match for godot 

This issue contains a lot of data about the problem imo - so it will be better to keep the request in it even with the noise. But if the devs want to, I can more it out to another issue and close this one

**Comment by DKesserich** [comment]
I don't believe OS.execute is working _at all_ on Windows. At least not in Godot 3.

No matter what command I pass into it it always returns a blank string. I'm trying to use it to run ffprobe on a selected video so I can get the number of audio tracks in the video, and it gives me nothing. I've tried having it launch cmd.exe and it still gives me nothing. It should at a minimum be giving me the info output that happens when you run cmd.exe in an open command prompt.

**Comment by brainsick** [comment]
I managed to scrape together a Windows 7 box that can run Godot, barely.

This works as of Godot 3.0 Beta 1

```
	var output = []
	var pid = OS.execute('cmd', ['/C', 'cd %TEMP% && chdir && echo Hello, Command Prompt!'], true, output)
	print('pid: ', pid, ' output: ', output)
```

Produces:

```
pid: 1069204049 output: [C:\Users\Todd\AppData\Local\Temp
Hello, Command Prompt!
]
```

**Comment by akien-mga** [comment]
Closing as the documentation now has a working example since #15917.

**Comment by rainlizard** [comment]
Here's how to set the working path. The help file example doesn't provide a very clear answer to the title of this issue.

``OS.execute('cmd', ['/C', 'start /D "F:\\Games\\Game Folder\\" game.exe'], false)``

**Comment by eon-s** [comment]
@rainlizard if documentation is still wrong/missing details, open an issue or make a PR with fixes in the documentation repository.
